gr-digital/lib/constellation.cc

```diff
@@ -304,26 +320,32 @@ std::vector<float> constellation::calc_soft_dec(gr_complex sample, float npwr)
 }
 
 void constellation::set_soft_dec_lut(const std::vector<std::vector<float>>& soft_dec_lut,
-                                     int precision)
+                                     int precision, float npwr)
 {
     max_min_axes();
 
     d_soft_dec_lut = soft_dec_lut;
     d_lut_precision = precision;
     d_lut_scale = powf(2.0, static_cast<float>(precision));
+    if (npwr != d_npwr && npwr != 1.0){
+        d_npwr = npwr;
+    }
 }
 
 bool constellation::has_soft_dec_lut() { return !d_soft_dec_lut.empty(); }
 
 std::vector<std::vector<float>> constellation::soft_dec_lut() { return d_soft_dec_lut; }
 
-std::vector<float> constellation::soft_decision_maker(gr_complex sample)
+std::vector<float> constellation::soft_decision_maker(gr_complex sample, float npwr)
 {
     if (has_soft_dec_lut()) {
         // Clip to just below 1 --> at 1, we can overflow the index
         // that will put us in the next row of the 2D LUT.
-        float xre = branchless_clip(sample.real(), 0.99);
-        float xim = branchless_clip(sample.imag(), 0.99);
+        // Since we are expecting inputs between +-1. We will scale to
+        // match the constellation+padding defined in the LUT
+        float ptscale = 2.0f*d_maxamp;
+        float xre = branchless_clip(sample.real()/ptscale, 1-1e-7);
+        float xim = branchless_clip(sample.imag()/ptscale, 1-1e-7);
```

Comment:
The comment is a bit misleading, since we're actually expecting inputs between `+-ptscale`.
How about rewriting the comment and the code around it to make it more clear?
```c++
        float xre = branchless_clip(sample.real(), ptscale-1e-7);
        float xim = branchless_clip(sample.imag(), ptscale-1e-7);
        // [...]
        float scale = d_lut_scale / (2.0f * ptscale);
        re = floorf((ptscale + xre) * scale);
        xim = floorf((ptscale + xim) * scale);
``` 

Also, I wonder if it would be better to use `roundf()` instead of `floorf()` here. `roundf()` seems better in the sense that it actually gets the nearest point from the LUT, but I haven't thought carefully if it comes with any nasty surprises such as an off-by-one error in the LUT index when we round up at the very end of the LUT (though this wouldn't be too problematic, because then there is a check that the index is in bounds).

Reply:
so this is also tricky. I think this whole LUT function is evil. this is the whole reason I went with my funny `ptscale` method. I originally intended to modify `maxd` and basically change the range that the indices took on. This proved to be very confusing and I was unsuccessful in getting it to work. Later however, I realized much of my trouble was because of some other bug with resolutions of 9-11 bits, and of course the whole time I was testing this was with 9 bits.
The `ptscale` scalar really just made understanding the indexing here easier for me, and I don't see why changing the index range shouldn't work as long as it's consistent to the generation of the LUT. However, I would really avoid messing with the `floorf()` because that's specifically done to take advantage of the indices being "overflowed" by having 1.0 added after clipping to push into the next index. It's very confusing, I hate it, and I don't want to mess with it now that it works.